### PR TITLE
A generator-runner developer guide

### PR DESCRIPTION
**Pull request by christian-monch**
This PR adds a first version of a developer guide to the generator-runner.

TODO:

- [x] add programming examples

### CODE REVIEW

**Comment by adswa** [comment]
Preview link: https://datalad--476.org.readthedocs.build/projects/next/en/476/developer_guide/generator-runner.html

File: docs/source/developer_guide/generator-runner.rst [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+- data received from stdout or stderr of the subprocess: ``pipe_data_received``
+- data connection lost: ``pipe_connection_lost``
+- connection to subprocess established: ``connection_made``
+- connection to subprocess closed: ``connection_lost``
+- subprocess exited: ``process_exited``
+- timeout occurred: ``timeout``
+
+When the subprocess is executed the runner will invoke the appropriate callbacks.
+
+``datalad_next`` provides some pre-defined protocol-classes for common use cases, e.g. ``datalad_next.runners.StdOutErrCapture``, which executes a subprocess and returns a dictionary with the ``stdout``- and ``stderr``-output, and the `return code` (also referred to as `exit status` ) of the subprocess.
```
Comment (adswa) [inline docs/source/developer_guide/generator-runner.rst:30]: This sounds like a good place to mention all pre-defined protocol classes to give a quick overview?

(also found a stray space)
```suggestion
``datalad_next`` provides some pre-defined protocol-classes for common use cases, e.g. ``datalad_next.runners.StdOutErrCapture``, which executes a subprocess and returns a dictionary with the ``stdout``- and ``stderr``-output, and the `return code` (also referred to as `exit status`) of the subprocess.
```

File: docs/source/developer_guide/generator-runner.rst [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+4. If the internal result queue of the generator is not empty, the generator will yield the first element of the result queue and go back to step 2.
+
+5. The subprocess has exited, the next call to the its ``send()``-method will retrieve the exit status of the subprocess and end the iteration by raising ``StopIteration``.
+
+From the description above and from the figure it should be clear, that, if a user wants to receive output from a subprocess and not just wait for its exit, the user has to send
+data that is received via ``pipe_data_received`` to the result queue of the result generator.
+That means basically he has to call ``send_result()``.
+This is not automatically mixed into the protocol class by ``GeneratorMixIn``.
+A minimal generator protocol would therefore look like this.
+
```
Comment (adswa) [inline docs/source/developer_guide/generator-runner.rst:100]: I'm not 100% sure, is this something that already exists in the code? Or is it something one would need to implement if this functionality is desired?

Comment (christian-monch) [reply docs/source/developer_guide/generator-runner.rst:100]: Yes, that is confusing. The outlined action, i.e. calling `send_result()` in the `pipe_data_received``-callback, is implemented in all provided generator-protocols. But if you create your own protocol implementation, you will have to do it as well.

I will try to make this more clear.

File: docs/source/developer_guide/generator-runner.rst (hunk 1 above)
Comment (christian-monch) [reply docs/source/developer_guide/generator-runner.rst]: > This sounds like a good place to mention all pre-defined protocol classes to give a quick overview?

That is a good idea. I will add that. Depending on how bulky it gets, I might just mention the "top" ("top" as in most-used)  protocols.

File: docs/source/developer_guide/generator-runner.rst (hunk 2 above)
Comment (christian-monch) [reply docs/source/developer_guide/generator-runner.rst]: Addressed in cd13ec7c13abeb5f10cc584ab7f48c7f13f1785c

File: docs/source/developer_guide/generator-runner.rst (hunk 1 above)
Comment (christian-monch) [reply docs/source/developer_guide/generator-runner.rst]: On  a second thought. At the position mentioned above the generator-mode is not yet described. I will therefore add the list at the end the section "General functionality", just before the section "Programming examples".

Comment (christian-monch) [reply docs/source/developer_guide/generator-runner.rst]: done in 43fc79a4e5cf3067fe04d1bd8e9c5787e8bb2a1d

File: docs/source/developer_guide/generator-runner.rst (hunk 2 above)
Comment (christian-monch) [reply docs/source/developer_guide/generator-runner.rst]: Done in cd13ec7c13abeb5f10cc584ab7f48c7f13f1785c

**Comment by mih** [comment]
If we follow through with #541, it would be suboptimal to include this documentation in datalad-next. I suggest to repost the PR against datalad-core. I believe that even if we stop using that runner implementation over here, it will continue to live for a long time in datalad-core.

I think a document in the set of design docs (or next to it) would be best.

**Pull request closed without merging**